File: 56.merge-intervals.cpp
/*
 * @lc app=leetcode id=56 lang=cpp
 *
 * [56] Merge Intervals
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty()) {
            return {};
        }
        sort(intervals.begin(),
             intervals.end(), 
             [](vector<int>& i1, vector<int>& i2){
                 return i1[0] < i2[0];
             });
        vector<vector<int>> result;
        for (int i = 0; i < intervals.size();) {
            int left = intervals[i][0];
            int cur_right = intervals[i][1];
            if (i == intervals.size() - 1) {
                result.push_back({left, cur_right});
                break;
            }
            for (int j = i + 1; j < intervals.size(); ++j) {
                int next_left = intervals[j][0];
                int next_right = intervals[j][1];
                // 如果两个区间完全不重叠
                if (next_left > cur_right) {
                    i = j;
                    result.push_back({left, cur_right});
                    break;
                } else {
                    // 如果一个区间包含另一个区间
                    if (next_right <= cur_right) {
                        // 不扩大区间
                    } else {
                        // 如果只是部分重叠
                        // 扩大区间右界
                        cur_right = next_right;
                    }
                    if (j == intervals.size() - 1) {
                        result.push_back({left, cur_right});
                        i = j + 1;
                        break;
                    }
                }
            }
        }
        return result;
        // list<vector<int>> il(intervals.begin(), intervals.end());
        // for (auto i = il.begin(); i != --il.end(); ) {
        //     vector<int>& left = *i;
        //     vector<int>& right = *++i;
        //     if (left[1] >= right[0]) {
        //         if (left[1] < right[1]) {
        //             left[1] = right[1];
        //         }
        //         i = --il.erase(i);
        //     } else {
        //         ++i;
        //     }
        // }
        // return vector<vector<int>>(il.begin(), il.end());
    }
};
// @lc code=end

File: 56.merge-intervals.java
import java.util.Arrays;

/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */

// @lc code=start
class Solution {
    public int[][] merge(int[][] intervals) {
        int intervalCount = intervals.length;
        final int n = intervals.length;
        // 先对区间起始位置进行排序
        Arrays.sort(intervals, (a, b)->{return a[0] - b[0];});
        for (int i = 0, k = 0; i < n; ++i, ++k) {
            // 当前要处理的区间i
            int[] interval = intervals[i];
            int back = interval[1];
            int j = i + 1;
            for (; j < n; ++j) {
                // 如果区间没有overlap
                if (back < intervals[j][0]) {
                    break;
                }
                // 如果有overlap，更新区间结束位置
                back = Math.max(back, intervals[j][1]);
                // 合并使区间数量减少一个
                intervalCount--;
            }
            // 放到实际应该保存的位置k
            intervals[k][0] = interval[0];
            intervals[k][1] = back;
            // 新的要检查区间的位置，由于i还有++，所以这里取j-1
            i = j - 1;
        }
        int[][] res = new int[intervalCount][2];
        for (int i = 0; i < intervalCount; ++i) {
            res[i][0] = intervals[i][0];
            res[i][1] = intervals[i][1];
        }
        return res;
    }
}
// @lc code=end

